app/multitenant/dynamo_collector.go

```diff
@@ -345,6 +408,72 @@ func (c *dynamoDBCollector) getReports(userid string, row int64, start, end time
 	return append(cachedReports, fetchedReport...), nil
 }
 
+func memcacheStatusCode(err error) string {
+	// See https://godoc.org/github.com/bradfitz/gomemcache/memcache#pkg-variables
+	switch err {
+	case nil:
+		return "200"
+	case memcache.ErrCacheMiss:
+		return "404"
+	case memcache.ErrMalformedKey:
+		return "400"
+	default:
+		return "500"
+	}
+}
+
+func (c *dynamoDBCollector) fetchFromMemcache(reportKeys []string) ([]report.Report, []string, error) {
+	var foundReports map[string]*memcache.Item
+	err := timeRequest("Get", memcacheRequestDuration, memcacheStatusCode, func() error {
+		var err error
+		foundReports, err = c.memcache.GetMulti(reportKeys)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, reportKeys, err
+	}
+
+	// Decode all the reports in parallel.
+	type result struct {
+		key    string
+		report *report.Report
+	}
+	ch := make(chan result, len(reportKeys))
```

Comment:
I tried and couldn't figure out how to do this without adding a bunch of extra book-keeping to track how many things we'd sent to the `missing` channel.

Have updated to use your preferred early-exit style.